lib/wasi/src/runners/wcgi/handler.rs

```diff
@@ -0,0 +1,235 @@
+use std::{
+    collections::HashMap,
+    path::{Path, PathBuf},
+    pin::Pin,
+    sync::Arc,
+    task::Poll,
+};
+
+use anyhow::Error;
+use futures::{Future, FutureExt, StreamExt, TryFutureExt};
+use http::{Request, Response};
+use hyper::{service::Service, Body};
+use tokio::{
+    io::{AsyncBufReadExt, AsyncRead, AsyncWrite, AsyncWriteExt},
+    runtime::Handle,
+};
+use wasmer::Module;
+use wasmer_vfs::{FileSystem, PassthruFileSystem, RootFileSystemBuilder, TmpFileSystem};
+use wcgi_host::CgiDialect;
+
+use crate::{
+    http::HttpClientCapabilityV1,
+    runners::wcgi::{Callbacks, MappedDirectory},
+    Capabilities, Pipe, PluggableRuntimeImplementation, VirtualTaskManager, WasiEnv,
+};
+
+/// The shared object that manages the instantiaion of WASI executables and
+/// communicating with them via the CGI protocol.
+#[derive(Clone, derivative::Derivative)]
+#[derivative(Debug)]
+pub(crate) struct Handler {
+    pub(crate) program: Arc<str>,
+    pub(crate) env: Arc<HashMap<String, String>>,
+    pub(crate) args: Arc<[String]>,
+    pub(crate) mapped_dirs: Arc<[MappedDirectory]>,
+    pub(crate) task_manager: Arc<dyn VirtualTaskManager>,
+    pub(crate) module: Module,
+    pub(crate) dialect: CgiDialect,
+    #[derivative(Debug = "ignore")]
+    pub(crate) callbacks: Arc<dyn Callbacks>,
```

Comment:
Having everything as an arc here seems a bit wrong.
Why not create an inner type and wrap just only that type inside the Arc?

Reply:
Done.